### ISSUE TITLE
Window freezes on resize in Windows 10 (Default & P2D renderers)

### ISSUE THREAD

**Issue by B3ret** [processing/processing#5579]
## Description
If I set the sketch window (of ANY sketch) to be resizeable and start resizing it using the mouse, at one point the render window just freezes for the Default renderer. I have to close the application and restart it.
The same thing happens, if I pick the P2D renderer instead; just the handles for resizing the window are missing.

If I pick the FX2D-renderer, the problem is gone. (However, then I have other problems, because surface.setSize on FX2D behaves weirdly.)

## Expected Behavior
No freeze should occur.

## Current Behavior
Window freezes with black areas at the resized portion.

## Steps to Reproduce
Any resize-able processing sketch will produce the problem with the above renderers, both from the Processing development environment and from exported jar-File.

0. Ensure you are on Windows 10. (About a month ago, I couldn't reproduce on Windows 7. I don't have Win 7 around right now to test it again.)
1. Minimal code example

```
void setup()
{
  size(800, 600);
  surface.setResizable(true);
}

void draw()
{
}
```

2. Grab the lower right corner with the mouse, and move it rapidly in circles. (Will also occur with normal resizing, just takes some time.)

3. Render window freezes

## Your Environment

 

* Processing version: 3.3.7
* Operating System and OS version: Windows 10, latest patches. 
* Other information: Lenovo T540p Geforce GT 730m (no clue if the dedicated card or the Intel Chipset is used for Processing though.)
* The problem appeared for at least one other Windows 10 machine, as discussed here: https://discourse.processing.org/t/windows-10-freeze-on-resizing/1813

## Possible Causes / Solutions
The bug started occurring on all processing versions and old exported application shortly after Windows 10 Spring update. Could be coincidence ofc.

**Comment by viv1kk** [comment]
mine is also freezing...

**Comment by liquid600pgm** [comment]
![](https://i.imgur.com/3YH3a2Q.gif)

I guess this is what happens.

This issue did not occur until I started recording that gif, so it may be a thing with slower processors/graphic cards.

**Comment by B3ret** [comment]
Hm. I'm actually on a quite fast machine, but yes. I really have to overdo the rapid Resizing to make it happen reliably.

FX2D would be fine for me, except of initial sizing on HighDPI: https://discourse.processing.org/t/set-initial-window-size-relative-to-screen-size/1814/3
Shall I create a bug report for this as well? Since displayHeight/displayWidth is deprecated, I wasn't sure if it's a bug or a feature...

**Comment by benfry** [comment]
Sounds like a JOGL problem with the Windows 10 update; someone will need to look into it. 

Glitchy things like this are usually threading issues, which tend to show up most acutely on very fast or very slow machines. 

@B3ret Yes, please file that issue separately.

**Comment by B3ret** [comment]
This is affecting a live product of mine. Can someone give me a TL;DR version for investigating this myself? (I'm a professional software developer, I'm just not into Java **at all**.)
I guess something like

1. Get JOGL from Git
2. Get Processing from Git
3. Setup Development/Debugging environment (Eclipse? Anything more lightweight? Any pointers on how to do this?)

**Comment by B3ret** [comment]
I don't want to be annoying, but can I get an update on this? Or an answer to my last question? I'd be willing to invest a few hours into debugging this.

**Comment by liquid600pgm** [comment]
@B3ret You don't need to clone JOGL, Processing will be enough.
I'd recommend IntelliJ IDEA as your IDE, as it has many integrated debugging tools, it is fast and very intelligent.
To set your dev environment up, clone the Processing git repository, and open IntelliJ IDEA. Go to File > Open, and navigate to your downloaded repo's root. Select the `core` folder. Now, Open the 'Ant Build' tab on the right side of your window, and drag and drop the `build.xml` file from the Project tab (left side).
To create a test sketch, create a package in the `src` folder (can be named `test` or anything), just remember to delete it before pushing. Here create a new class (for example `DebuggingSketch`), and paste in the following code:
```java
import processing.core.*;

public class DebuggingSketch extends PApplet implements PConstants {
  public static void main(String[] args) {
    PApplet.main(DebuggingSketch.class);
  }

  public void settings() {
    size(800, 600, P2D);
  }

  public void setup() {

  }

  public void draw() {

  }
}
```
and modify it to your will. Of course, don't forget to change the name of the class in your code, you can do it using Shift + F6 to Rename.

**Comment by lesto** [comment]
i have the same issue with program like hterm and explorer.exe itself.. so there may be a more fundamental issue with win 10

**Comment by B3ret** [comment]
> i have the same issue with program like hterm and explorer.exe itself.. so there may be a more fundamental issue with win 10

I do **not** have this problem with other software. Neither with explorer, nor with hterm (which I downloaded to test this).

**Comment by lesto** [comment]
sorry i must point out i got this issue only on the second monitor, that i use in vertical orientation.

**Comment by B3ret** [comment]
This problem has gone away for me. I guess it was a Windows 10 problem and they fixed it.
I can confirm this even for older Processing versions, since I have compiled versions of my Software and checked also the older ones.

Someone else who could reproduce the issue, please check again and maybe close the issue.

**Comment by jhwblender** [comment]
Problem has not gone away for me

**Comment by gouessej** [comment]
@B3ret Do you use a virus scanner? Have you ever installed Avira or Avast on your machine?

**Comment by B3ret** [comment]
> 
> 
> @B3ret @jhwblender Do you use a virus scanner? Have you ever installed Avira or Avast on your machine?

No. Just Windows defender.

**Comment by JakubValtar** [comment]
The default renderer runs in a background animation thread, not EDT, and there is no synchronization between the two. This means that any interaction between the sketch and AWT has a potential to cause threading issues. 

Calls to `surface` methods from the sketch are unsafe. Resizing is also unsafe, because AWT is changing the size of the canvas on EDT, while the sketch is drawing on it in the background thread. This is not an issue with your sketches, it is something we should fix.

I can't focus on this issue right now, but if somebody picks it up, the easy solution would be to run the default renderer sketches on the EDT (downside is that slow sketches would make the window unresponsive), the difficult solution would be to synchronize all the interactions between the EDT and the animation thread (PSurfaceAWT is the place to do it) without deadlocking each other. I can't be sure this is the cause of the problem, but there is a good chance it will fix it.

**Comment by B3ret** [comment]
As stated above, I don't have this issue anymore and it is also fixed for old versions of Processing. So I assume a Windows update fixed it.

**Comment by x-e-r-o** [comment]
Having this issue currently.

**Comment by benfry** [comment]
This should be resolved in 4.0 beta 5.
